### PR TITLE
feat(db): add prisma as ORM

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

There's nothing fancy here, I just followed

- https://www.prisma.io/fastify
- https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project/mongodb-typescript-mongodb

The schema.prisma was generated by `npx prisma db pull` with no effort on my part to clean it up. The comments (created by prisma) highlight where it could not nail down a specific data type.

### CODE REVIEW

File: api/prisma/schema.prisma [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  when String   @map("_when")
+}
+
+type UserBadges {
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  coreTeam Json?
+}
+
+type UserCompletedChallenges {
+  completedDate Float
```
Comment (ShaunSHamilton) [inline api/prisma/schema.prisma:21]: This is a float???

File: api/prisma/schema.prisma [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  userId  String   @db.ObjectId
+}
+
+model expressRateRecords {
+  id             String @id @default(auto()) @map("_id") @db.ObjectId
+  /// Field referred in an index, but found no data to define the type.
+  expirationDate Json?
+
+  @@index([expirationDate], map: "expirationDate_1")
+}
```
Comment (ShaunSHamilton) [inline api/prisma/schema.prisma:85]: 1) Does this have to do with our rate limiting?
2) If so, do we store this in a collection on MongoDB?
3) If so, should we not be using a volatile storage option for this?
4) This is not a permanent ban, is it?

**Comment by ojeytonwilliams** [comment]
To run it locally you might need to
```
cd api
npm run prisma generate
```

File: api/prisma/schema.prisma (hunk 2 above)
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:85]: 1. Yes
2. Yes
3. That would make more sense, I think.
4. Nope, hence the expiration date.

File: api/prisma/schema.prisma (hunk 1 above)
Comment (ShaunSHamilton) [reply api/prisma/schema.prisma:21]: Oh. This is just JavaScript being JavaScript - numbers are either Big or DP Floats

File: api/prisma/schema.prisma (hunk 2 above)
Comment (ShaunSHamilton) [reply api/prisma/schema.prisma:85]: I will create a pin for the new api

File: api/prisma/schema.prisma (hunk 1 above)
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:21]: I think it's simpler than that. We could have used Date (Mongo supports that after all), but we didn't.

File: api/prisma/schema.prisma [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  isRespWebDesignCert        Boolean
+  isSciCompPyCertV7          Boolean
+  keyboardShortcuts          Boolean?
+  location                   String
+  name                       String
+  picture                    String
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  portfolio                  Json?
+  profileUI                  UserProfileUi
+  progressTimestamps         Float[]
```
Comment (Nirajn2311) [inline api/prisma/schema.prisma:136]: What exactly is this key for?

File: api/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+type UserBadges {
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  coreTeam Json?
+}
+
+type UserCompletedChallenges {
+  completedDate Float
+  id            String
+}
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply api/prisma/schema.prisma:23]: We discussed this a bit on Chat. This PR will serve to give us an idea of what the DX would be like.

If we move ahead with this, we would need to get the schemas from the production database.

[the remark this answers is not in the document]
Comment (Nirajn2311) [reply api/prisma/schema.prisma:23]: Gotcha, I'll resolve the other comments then

File: api/prisma/schema.prisma [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  completedChallenges        UserCompletedChallenges[]
+  currentChallengeId         String
+  email                      String
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  emailAuthLinkTTL           Json?
+  emailVerified              Boolean
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  emailVerifyTTL             Json?
+  is2018DataVisCert          Boolean
+  is2018FullStackCert        Boolean
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:109]: Are you sure? It appears in the user model:

https://github.com/freeCodeCamp/freeCodeCamp/blob/310662fc808f780259ca186e94ccb999d8768bba/api-server/src/common/models/user.json#L194-L198

File: api/prisma/schema.prisma [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  isApisMicroservicesCert    Boolean
+  isBackEndCert              Boolean
+  isBanned                   Boolean
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:112]: Again, it's in the user.json schema.

File: api/prisma/schema.prisma (hunk 5 above)
[the remark this answers is not in the document]
Comment (Nirajn2311) [reply api/prisma/schema.prisma:109]: Both user objects of my account in staging and prod site don't have this key in them. Maybe it was added later on when we migrated to the latest step-by-step curriculum

[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:109]: Yep, sorry, you're right. The loopback docs confirm that all properties are optional by default

File: api/prisma/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  isCheater                  Boolean
+  isDataAnalysisPyCertV7     Boolean
+  isDataVisCert              Boolean
+  isDonating                 Boolean
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply api/prisma/schema.prisma:116]: We should just look at what MongoDB Atlas says on the production DB about how many schemas we have

[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/prisma/schema.prisma:116]: Can we just get the entire schema directly?  That would remove all the guesswork.

**Comment by ojeytonwilliams** [comment]
I've gone back and forth on this, but I think we should fix the schema in a followup PR. There's already plenty to QA without even thinking about the shape of the data.

File: api/prisma/schema.prisma (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply api/prisma/schema.prisma:116]: We can. I am busy testing on staging to see if it is a heavy operation.

File: api/prisma/schema.prisma (hunk 3 above)
Comment (Nirajn2311) [reply api/prisma/schema.prisma:136]: Gotcha thanks.

**Review by raisedadead — APPROVED** [review]

**Review by ShaunSHamilton — APPROVED** [review]
Works 👍

**Pull request merged**